### PR TITLE
#2908 Provide Doc Action voiding for contract period 

### PR DESCRIPTION
**Pull request by cristinamghita**
#2908 Provide Doc Action voiding for contract period

### CODE REVIEW

File: de.metas.contracts/src/main/java/de/metas/contracts/flatrate/process/C_Flatrate_Term_Change.java [hunk 1]
```diff
@@ -33,6 +32,8 @@
 
 	public static final String PARAM_TERMINATION_MEMO = I_C_Flatrate_Term.COLUMNNAME_TerminationMemo;
 	public static final String PARAM_TERMINATION_REASON = I_C_Flatrate_Term.COLUMNNAME_TerminationReason;
+
```
Comment (metas-ts) [inline de.metas.contracts/src/main/java/de/metas/contracts/flatrate/process/C_Flatrate_Term_Change.java]: please update your save actions like this:
image

goal: reduce meaningless differences for review etc

File: de.metas.contracts/src/test/java/de/metas/contracts/impl/ContractChangeBLTest.java [hunk 2]
```diff
@@ -164,5 +191,20 @@ private void assertSubscriptionProgress(@NonNull final I_C_Flatrate_Term flatrat
 				.filter(progress -> progress.getEventDate().after(flatrateTerm.getMasterEndDate()))
 				.peek(progress -> assertThat(progress.getContractStatus()).isEqualTo(X_C_SubscriptionProgress.CONTRACTSTATUS_Running));
 	}
+	
+	private void assertClosedFlatrateTerm(@NonNull final I_C_Flatrate_Term flatrateTerm)
+	{
+		assertThat(flatrateTerm.getDocStatus()).isEqualTo(X_C_Flatrate_Term.DOCSTATUS_Completed);
```
Comment (metas-ts) [inline de.metas.contracts/src/test/java/de/metas/contracts/impl/ContractChangeBLTest.java]: shouldn't the term have docstaus *closed* afterwards?

File: de.metas.contracts/src/test/java/de/metas/contracts/impl/ContractChangeBLTest.java [hunk 3]
```diff
@@ -131,6 +133,31 @@ public void cancel_a_Contract_which_was_extended_using_a_date_from_extended_cont
 		assertSubscriptionProgress(extendedContract, 3);
 		assertThat(contract.getMasterEndDate()).isEqualTo(extendedContract.getMasterEndDate());
 	}
+	
+	@Test
+	public void closeContract_test()
+	{
```
Comment (metas-ts) [inline de.metas.contracts/src/test/java/de/metas/contracts/impl/ContractChangeBLTest.java]: * this test seems to be about *not* closing the term, right? if so, pls update the method name
* and imho you can remote the `_test`, it's redundant

File: de.metas.contracts/src/test/java/de/metas/contracts/impl/ContractChangeBLTest.java [hunk 4]
```diff
@@ -131,6 +133,31 @@ public void cancel_a_Contract_which_was_extended_using_a_date_from_extended_cont
...
+		final ContractChangeParameters contractChangeParameters = ContractChangeParameters.builder()
+				.changeDate(SystemTime.asDayTimestamp())
+				.isCloseInvoiceCandidate(true)
+				.isOnlyTerminateCurrentTerm(true)
+				.build();
+		
+		assertThatThrownBy(() -> {
+			contractChangeBL.cancelContract(contract, contractChangeParameters);
+		}).hasMessageContaining(ContractChangeBL.MSG_IS_NOT_ALLOWED_TO_TERMINATE_CURRENT_CONTRACT);
+
```
Comment (metas-ts) [inline de.metas.contracts/src/test/java/de/metas/contracts/impl/ContractChangeBLTest.java]: aaaah, now I figured you are testing two things.
please do two distinct tests for them..ofc you can extract the setup code into a common method..

**Comment by metas-ts** [comment]
@cristinamghita 
Aaargh, sorry, i "accidentally" integrated..need coffee....please check out my [review](https://github.com/metasfresh/metasfresh/pull/2983#pullrequestreview-77023510) and make the fixes e.g. in a branch like `gh2908_1` or something

File: de.metas.contracts/src/test/java/de/metas/contracts/impl/ContractChangeBLTest.java (hunk 2 above)
Comment (cristinamghita) [reply de.metas.contracts/src/test/java/de/metas/contracts/impl/ContractChangeBLTest.java]: No, I still completed. Just the ConstractStatus is Quit. In the end, we using the same process as for general cancelling.

**Pull request merged**